pgdedupe/mssql/cli.py

```diff
@@ -0,0 +1,526 @@
+# -*- coding: utf-8 -*-
+
+"""
+Based on: https://github.com/datamade/dedupe-examples/tree/master/pgsql_big_dedupe_example
+"""
+import collections
+import csv
+import logging
+import os
+import random
+import sys
+import tempfile
+import time
+
+import dedupe
+
+import numpy
+
+import pymssql
+
+sys.path.append(os.path.abspath('pgdedupe'))
+from pgdedupe import exact_matches
+
+START_TIME = time.time()
+PYTHON_VERSION = sys.version_info[0]
+
+
+def mssql_main(config, db):
+    con = pymssql.connect(**db)
+    config = process_options(config)
+    config['database'] = db['database']
+
+    logging.info("Preprocessing...")
+    preprocess(con, config)
+
+    logging.info("Training...")
+    deduper = train(con, config)
+
+    logging.info("Creating blocking table...")
+    create_blocking(deduper, con, config)
+
+    logging.info("Clustering...")
+    clustered_dupes = cluster(deduper, con, config)
+
+    logging.info("Writing results...")
+    write_results(clustered_dupes, con, config)
+
+    logging.info("Applying results...")
+    apply_results(con, config)
+
+    # Close our database connection
+    con.close()
+
+    print('ran in', time.time() - START_TIME, 'seconds')
+
+
+def process_options(c):
+    config = dict()
+    # Required fields
+    for k in ('schema', 'table', 'key', 'fields'):
+        if k not in c:
+            raise Exception('Key ' + k + ' must be defined in the config file')
+        config[k] = c[k]
+    # Optional fields
+    for k, default in (('interactions', []),
+                       ('threshold', 0.5),
+                       ('recall', 0.90),
+                       ('merge_exact', []),
+                       ('settings_file', 'dedup_mssql_settings'),
+                       ('training_file', 'dedup_mssql_training.json'),
+                       ('filter_condition', '1=1'),
+                       ('num_cores', None),
+                       ('use_saved_model', False),
+                       ('prompt_for_labels', True),
+                       ('seed', None)
+                       ):
+        config[k] = c.get(k, default)
+    # Ensure that the merge_exact list is a list of lists
+    if type(config['merge_exact']) is not list:
+        raise Exception('merge_exact must be a list of columns')
+    if len(config['merge_exact']) > 0 and type(config['merge_exact'][0]) is not list:
+        config['merge_exact'] = [config['merge_exact']]
+    # Add variable names to the field definitions, defaulting to the field
+    for d in config['fields']:
+        if 'variable name' not in d:
+            d['variable name'] = d['field']
+    # Add some handy computed values for convenience
+    config['all_fields'] = config['fields'] + [
+        {'type': 'Interaction', 'interaction variables': x} for x in config['interactions']]
+    columns = set([x['field'] for x in config['fields']])
+    # Used to nested sub-queries
+
+    config['stuff_condition'] = ' AND '.join(
```

Comment:
I was using it before but changed it after using binary checksum.
Will remove it.